packages/server/src/http/buildResponse.ts

```diff
@@ -0,0 +1,310 @@
+      if (prop === 'text') return () => response?.text();
+      if (prop === 'json') return () => response?.json();
+      if (prop === 'body') return response.body;
+      return Reflect.get(target, prop);
+    },
+  });
+
+  return headProxy;
+
+  async function* createBodyGenerator(
+    head: StreamHTTPResponse,
+  ): AsyncGenerator<string, void, number> {
+    try {
+      const formatOutput = createBodyFormatter({
+        style,
+        router,
+        head,
+        onResponseInit: ({ data = [], errors = [] } = {}) =>
+          responseMeta?.({
+            data,
+            errors,
+            ctx,
+            paths,
+            type,
+            eagerGeneration: style.includes('stream'),
+          }) ?? {},
+      });
+
+      // Map each promise to a tuple of its index and its result so we can resolve them out of order and still
+      // associate them with the correct procedure call
+      const indexedPromises = new Map(
+        promises.map((promise, index) => [
+          index,
+          promise.then((r) => [index, r] as const),
+        ]),
+      );
+
+      // Start counting the chunks at the last index of results to avoid key collisions
+      let chunkIndex = indexedPromises.size;
+
+      async function getNextPromise() {
+        const [index, value] = style.includes('stream')
+          ? await Promise.race(indexedPromises.values())
+          : await Array.from(indexedPromises.values())[0]!;
+        indexedPromises.delete(index);
+        return [index, value] as const;
+      }
+
+      while (indexedPromises.size > 0) {
+        const [index, envelope] = await getNextPromise();
+        // The result is an error, so return it as a complete procedure result
+        if ('error' in envelope) {
+          yield formatOutput(index, envelope);
+          continue;
+        }
+
+        // If the procedure returned a function assume it's a generator fn and call it to get the generator
+        envelope.result.data =
+          typeof envelope.result.data === 'function'
+            ? envelope.result.data()
+            : envelope.result.data;
+
+        // If the data is now an async iterable, call its async iterator method to get the async iterator
+        if (isAsyncIterableEsque(envelope.result.data)) {
+          envelope.result.data = envelope.result.data[Symbol.asyncIterator]();
+        }
+
+        // If we don't have an async iterator at this point, we can't stream it, so just return the result
+        if (!isAsyncIteratorEsque(envelope.result.data)) {
+          yield formatOutput(index, envelope);
+          continue;
+        }
+
+        const chunks = getRemainingChunks(envelope.result.data);
+        //XXX
+        if (style !== 'event-stream' && style !== 'json-stream') {
+          // We're not streaming, so we need to resolve the generator and return the complete result
+          let done;
+          let value;
+
+          while (!done) ({ done, value } = await chunks.next());
```

Comment:
I think we should just throw here instead and have the rule if you request an async generator, you have to use a transport/http link that supports it

Reply:
It seems intuitive that returning a generator from a tRPC resolver wouldn't work without additional configuration. I guess my thought was preserving compatibility in the case you have multiple clients in heterogenous runtimes with varying degrees of streaming support, e.g. mobile and web.

My toy use-case for the SSE side of things was the GPT4 API, but blocking on a mobile device until the response completes is a pretty crap experience in that case, so it's probably an unrealistic edge-case unless there's some other scenario where blocking for an otherwise-streamed response is acceptable UX.